components/fonts/shapers/mod.rs

```diff
@@ -0,0 +1,237 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+mod harfbuzz;
+use std::cmp;
+
+use app_units::Au;
+use base::text::is_bidi_control;
+use euclid::default::Point2D;
+use fonts_traits::ByteIndex;
+pub use harfbuzz::{ShapedGlyphData, Shaper};
+use log::debug;
+use num_traits::Zero as _;
+
+const NO_GLYPH: i32 = -1;
+
+use crate::{
+    Font, FontBaseline, GlyphData, GlyphId, GlyphStore, ShapingOptions, advance_for_shaped_glyph,
+};
+
+/// Utility function to convert a `unicode_script::Script` enum into the corresponding `c_uint` tag that
+/// harfbuzz uses to represent unicode scipts.
+fn unicode_script_to_iso15924_tag(script: unicode_script::Script) -> u32 {
+    let bytes: [u8; 4] = match script {
+        unicode_script::Script::Unknown => *b"Zzzz",
+        _ => {
+            let short_name = script.short_name();
+            short_name.as_bytes().try_into().unwrap()
+        },
+    };
+
+    u32::from_be_bytes(bytes)
+}
+
+struct ShapedGlyphEntry {
+    codepoint: GlyphId,
+    advance: Au,
+    offset: Option<Point2D<Au>>,
+}
+
+/// A servo shaper
+pub(crate) trait TShaper: Send + Sync {
```

Comment:
Let's just stick with `HarfBuzzShaperTrait` for now and save the more generic `Shaper` for later. I don't think we need more levels of abstraction (3) than implementations (2). Perhaps you have in mind a non-HarfBuzz shaper in the future. That's fine, but let's wait to add this level of abstraction until we are doing that.

My hope is that in a bit when HarfRust has reached parity with HarfBuzz, we can just switch to using HarfRust for everything and remove the traits.